packages/back-end/src/routers/tag/tag.controller.ts

```diff
@@ -33,15 +44,92 @@ export const postTag = async (
   if (!context.permissions.canCreateAndUpdateTag()) {
     context.permissions.throwPermissionError();
   }
-  const { id, color, description } = req.body;
+  const { id, color, description, label } = req.body;
 
-  await addTag(context.org.id, id, color, description);
+  // make sure it doesn't already exist:
+  const existing = await getAllTags(context.org.id);
+  const matchingId = existing.find((tag) => tag.id === id);
+  const matchingLabel = existing.find((tag) => tag.label === label);
+
+  if (matchingId) {
+    throw new Error("A Tag with this id already exists");
+  }
+  if (matchingLabel) {
+    throw new Error("A Tag with this name already exists");
+  }
+  // add the tag:
+  await addTag(context.org.id, id, color, description, label);
+
+  // audit log:
+  await req.audit({
+    event: "tag.create",
+    entity: {
+      object: "organization",
+      id: context.org.id,
+    },
+    details: auditDetailsCreate({ id, color, description, label }),
+  });
 
   res.status(200).json({
     status: 200,
   });
 };
 
+type PutTagResponse = {
+  status: 200;
+};
+
+/**
+ * PUT /tag/:id
+ * Update a tag by id
+ * @param req
+ * @param res
+ */
+export const putTag = async (
+  req: AuthRequest<TagInterface, { id: string }>,
+  res: Response<PutTagResponse>
+) => {
+  const context = getContextFromReq(req);
+  if (!context.permissions.canCreateAndUpdateTag()) {
+    context.permissions.throwPermissionError();
+  }
+  const { label, color, description } = req.body;
+  const { id } = req.params;
+  const existing = await getAllTags(context.org.id);
+  const matchingId = existing.find((tag) => tag.id === id);
+  const matchingLabel = existing.find((tag) => tag.label === label);
+  if (!matchingId) {
+    throw new Error("Tag not found");
+  }
+  if (matchingLabel) {
+    throw new Error("A Tag with this name already exists");
+  }
```

Comment:
What if we are updating the same tag with the same label but changing `color`? I don't think that this should fail.